openhtf/io/frontend/src/app/test.ts

```diff
@@ -0,0 +1,191 @@
+// Copyright 2016 Google Inc.All Rights Reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+/** Represents an active or past test run. **/
+export class Test {
+  dutID: string;
+  startTime: number;
+  endTime: number;
+  duration: number;
+  outcome: string;
+  phases: Phase[];
+  logEntries: LogEntry[];
+
+  /**
+   * Create a Test.
+   * @param record - A serialized OpenHTF TestRecord.
+   */
+  constructor() {
+    this.duration = 0;
+    this.logEntries = [];
+    this.phases = [];
+  }
+
+  /**
+   * Update state based on a test record.
+   * @param test - A serialized OpenHTF TestState.
+   * @param phaseList - An option list of phases to override the record.
+   */
+  update(record: any) {
+    // Test state is tricky to track and update particularly because we want
+    // log entries and other data displayed in the gui to be selectable. If
+    // new components are being created on every update, we will lost active
+    // selections even when the gui appears to remain static. Instead, we need
+    // to keep the gui _actually_ static when no changes are happening, which
+    // means the inputs to view componts need to stay pointing at the same
+    // objects in memory. If someone has a better way to do this, please let
```

Comment:
At first, I was hesitant, but if you note that we can have tests with large attachments in the record and that we store them all in memory, then it makes sense. Every time you get a new state, would you generate the patch, throw out the new state, update your stored t-N state to t-N+1 and throw out the oldest patch, too? Seems like a lot of complexity?

The old system, while a tad over-engineered, essentially did this:

Frontend Server: Hold 'now' state with N patches for converting any state after now-N to now. When the state changes, generate a patch and send that to any connected websockets. Store the new state and patch, drop the old state and oldest patch.
UI: Hold 'now' state with timestamp (version number). Listen on websocket for patches. On reconnect, send request for 'next' state with timestamp of 'now', and receive either a list of patches or a brand-new state if the timestamp was from > now-N.

If we ignore the operational transform stuff (unnecessary when only 1 system can change the state at a time), then the system was pretty nice. Some tweaks had to be implemented to save bandwidth/latency, such as not sending attachments over the wire until requested, but I think those are relevant here as well, unfortunately.